Sprint 35 TLT-2086 Selenium tests for canvas course info tool

rascalking opened the pull request.

Commit:
wip on selenium for canvas_course_info

eparker71 committed:
This is still a work in progress, I got one test to work but it doesn't work with the new page locator system. I used the older method to locate an ID and add it to the test.

eparker71 committed:
made some changes to the tests, this is still a work in progress.

eparker71 committed:
fixed typo

Commit:
make using htmlrunner a secure.py setting

eparker71 committed:
added one test to check if the tool loads properly

Commit:
fix locator metaclass

Commit:
get basic page load test working

elliottyates committed:
enables content window to notify parent iframe of embedded document size via iframeResizer.js jquery plugin; also removes unused library files

elliottyates committed:
course info widget defaults to 200px height so that placeholder doesn't take up so much space; course info widget box on syllabus/wiki page now takes up full width of container to save page space

Commit:
add test that inserts the widget in the page

* move driver insantiation/teardown to per-test, rather than per-class
* add passthrough on base page to switch_to.default_content()

Commit:
verify the fields used for the widget

Commit:
verify that saved page contains widget

elliottyates commented on selenium_tests/regression_suite.py:
```diff
@@ -0,0 +1,33 @@
+> export PYTHONPATH=~/tlt/canvas_course_info
+> export DJANGO_SETTINGS_MODULE=canvas_course_info.settings.local
+> sudo apt-get install xvfb
+> python selenium_tests/regression_tests.py
```
`regression_suite.py`

elliottyates commented on selenium_tests/regression_suite.py:
```diff
@@ -0,0 +1,33 @@
+    )
+else:
+    import logging; logging.basicConfig(level=logging.DEBUG)
+    import unittest
```
This needs to be loaded after e.g. `import time` so that the `TestLoader()` on line 29 can be instantiated.

rascalking replied:
Good catch.

elliottyates commented on selenium_tests/base_page.py:
```diff
@@ -0,0 +1,127 @@
+    up the element found by SomePageLocators.TITLE, and SomePage.button,
+    returning the element found by SomePageLocators.BUTTON.
+    """
+    def __new__(class_, name, bases, dict_):
```
Can we write a unit test for this? :)
...
_(that's a joke)_

rascalking replied:
It's not a bad idea, actually...put it on the todo list for when we split this out into a separate project.

elliottyates commented on selenium_tests/course_info/course_info_base_test_case.py:
```diff
@@ -0,0 +1,19 @@
+
+class CourseInfoBaseTestCase(BaseSeleniumTestCase):
+    """
+    Lecture Video base test case, all other tests will subclass this class
```
Update comment

elliottyates commented on selenium_tests/course_info/course_info_base_test_case.py:
```diff
@@ -0,0 +1,19 @@
+        super(CourseInfoBaseTestCase, cls).setUpClass()
+        cls.USERNAME = settings.SELENIUM_CONFIG.get('selenium_username')
+        cls.PASSWORD = settings.SELENIUM_CONFIG.get('selenium_password')
+        cls.BASE_URL = '%s/courses/7162/pages/course-information/edit' % settings.SELENIUM_CONFIG.get('canvas_base_url')
```
Can we put this in the SELENIUM_CONFIG, like it's done in lti_emailer? Cf https://github.com/Harvard-University-iCommons/lti_emailer/blob/develop/lti_emailer/settings/local.py#L25

rascalking replied:
Yup.

elliottyates commented on selenium_tests/course_info/course_info_tests.py:
```diff
@@ -0,0 +1,106 @@
+from selenium_tests.pin.page_objects.pin_login_page_object import PinLoginPageObject
+
+
+class CourseInfoTestFlow(CourseInfoBaseTestCase):
```
Can we change this to just `CourseInfoTests`? It's not really a flow, and the tests should be mutually exclusive.

rascalking replied:
Done.

elliottyates commented on selenium_tests/course_info/course_info_tests.py:
```diff
@@ -0,0 +1,106 @@
+        edit_page = EditorPage(self.driver)
+
+        # delete any instances of the widget already in the page
+        edit_page.focus_on_default_content()
```
I don't know how much slower it is, but I think it reads better if these are part of the `PageObject`s... e.g. if something fails, check that we're in the iframe, and if not, switch to default content then switch to iframe. Seems like the test shouldn't need to think about which frame it's in.

rascalking replied:
Yeah, I'd been trying to avoid having methods on the page objects...with the located properties, it was originally very straightforward and easy to read without them.  But getting everything to work correctly required a lot of paranoia about where selenium was focusing, and I should have pushed that down.  I'm not a fan of "do the thing we're pretty sure will fail, then try to recover from it" as a pattern, though...I'm just going to move the explicit focus direction into methods.

(also, there's no way in selenium to find the name of the current frame you're in...one of many limitations that made me want to flip tables)

elliottyates commented on selenium_tests/course_info/course_info_tests.py:
```diff
@@ -0,0 +1,106 @@
+        pin_page.login(self.BASE_URL, self.USERNAME, self.PASSWORD)
+        edit_page = EditorPage(self.driver)
+
+        # delete any instances of the widget already in the page
```
This whole chunk of setting up and removing the existing widgets might be better in a private helper method; the test would be shorter and more focused on the 'insert'ing functionality.

rascalking replied:
Agreed.

elliottyates commented on selenium_tests/course_info/course_info_tests.py:
```diff
@@ -0,0 +1,106 @@
+        self.assertEqual(len(widgets), 1)
+
+        # figure out which f values we should be expecting
+        expected_f_values = {
```
Can we make this chunk a `PageObject` function, to encapsulate explicit references to its locator_class, etc?

rascalking replied:
Yup.

elliottyates commented on selenium_tests/course_info/course_info_tests.py:
```diff
@@ -0,0 +1,106 @@
+        self.assertEqual(edit_page.get_inserted_widgets(), [],
+                         'Found unexpected instances of widget already in the page')
+
+        # launch the tool, click save in the tool iframe
```
If this + the verify step were encapsulated in a `PageObject` function I think the test would be cleaner, and we could reuse the same series of steps more easily in another test.

rascalking replied:
I don't think we should be moving any verification/assert logic into the page objects.  But just like the other ones, I definitely agree this "do some focus logic and X" chunk belongs in a method.

elliottyates commented on selenium_tests/base_page.py:
```diff
@@ -0,0 +1,127 @@
+
+        class SomePage(BasePage):
+            locator_class = SomePageLocators
+            located_properties = ['title', 'button']
```
Any reason this isn't `locator_properties`? To me it reads like 'the list of properties that we located' vs. 'the list of locators that can be accessed as a class property'.

rascalking replied:
I'd intended it as "the list of properties we create using locators", actually.  I think when we pull this into a separate project, I'm going to do away with this altogether (too easy to forget to add/remove things from it) in favor of using two Locator classes (one for properties, referenced from `locator_class`, another for the non-property locators).  For now, I'm inclined to leave this alone.

elliottyates commented:
Cool. :+1: I like the elegance of the metaclass approach, but I'm sad that it means that IDE introspection isn't as straightforward anymore. (e.g. `self.editor_tabs` isn't an attribute the IDE knows about prior to run time, so shows up as a warning; for those less familiar with Python it may be confusing where that property comes from in the first place; cf https://github.com/Harvard-ATG/canvas_course_info/pull/13/files#diff-f62481f077a51c49911bbd9ae35f7101R66)

Commit:
updates per PR

* moves tool url path into settings
* moves (as much as possible) selenium focus commands out of tests and
  into page objects
* moves chunks of page-related logic into page object methods
* mess with PYTHONPATH/DJANGO_SETTINGS_MODULE from regression_suite.py
  so we don't have to do it by hand from the calling environment
* use test discovery so we don't have to manually import TestCase
  subclasses and then construct a TestSuite by hand.
